sql: test EXTRACT more and add EXTRACT_DURATION for INTERVAL

knz opened the pull request:
EXTRACT_DURATION for INTERVAL is a CockroachDB-specific feature.
It can only extract hours, minutes, seconds, milliseconds, microseconds.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9734)

<!-- Reviewable:end -->

knz committed:
sql: test EXTRACT more and add new EXTRACT_DURATION for INTERVAL.

EXTRACT_DURATION for INTERVAL is a CockroachDB-specific feature.
It can only extract hours, minutes, seconds, milliseconds, microseconds.

This patch also tests that the 2nd argument to
EXTRACT/EXTRACT_DURATION can be autocasted from string.

mjibson commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[sql/parser/eval_test.go, line 492 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9734#-KTGTRLJS6qgH6z_tDXk:-KTGTRLJS6qgH6z_tDXl:bkc4uhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/56a5bee3d59523b0421de5b8d62b76581d4c8db2/sql/parser/eval_test.go#L492)):*
> ```Go
> 		{`ANNOTATE_TYPE(NULL, timestamp)`, `NULL`},
> 		// Extract on timestamps and intervals.
> 		{`extract(year from '2010-09-28 12:13:14.1+00:00'::timestamp)`, `2010`},
> ```

Did this syntax previously work but now won't because of the additional override?

`extract(year from '2010-01-01 00:00:00')`

That is, the auto string-to-timestamp conversion?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9734)*
<!-- Sent from Reviewable.io -->

knz commented:
Renamed the new built-in function and added the missing test. PTAL

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/parser/eval_test.go, line 492 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9734#-KTGTRLJS6qgH6z_tDXk:-KTGVR9FZmPnreVEypo8:b-kak2hd) ([raw file](https://github.com/cockroachdb/cockroach/blob/56a5bee3d59523b0421de5b8d62b76581d4c8db2/sql/parser/eval_test.go#L492)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Did this syntax previously work but now won't because of the additional override?
> 
> `extract(year from '2010-01-01 00:00:00')`
> 
> That is, the auto string-to-timestamp conversion?
</details>
Argh, yes. I didn't see it because there was no test for it :(
There is now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9734#-:-KTGWcDeOdVQIio80s2M:b4s8ilr)*
<!-- Sent from Reviewable.io -->

mjibson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9734#-:-KTGZxoDTwYU4oMtlgmf:bnfp4nl)*
<!-- Sent from Reviewable.io -->
